### PR TITLE
Use system dynamic linker

### PR DESCRIPTION
**Pull request by tmcdonell**
**Description**
Changes to (hopefully) support compilation on Apple silicon Macs, together with AccelerateHS/accelerate#513.

Rather than load object (.o) files directly using our own runtime linker, we now convert that object file to a shared object file (.so/.dylib/.dll) and use the system linker. This results in more clutter in the cache (both a .o file to support ahead-of-time compilation as well as the .so for runtime compilation need to be stored) as well as some fuss to work around (presumably) a bug/limitation in how ghci loads dependent libraries (or rather, doesn't).

@robbert-vdh 

**Motivation and context**
AccelerateHS/accelerate-llvm#80

**How has this been tested?**
nofib and manual tests in ghci.

**Types of changes**

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Checklist:**

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: accelerate-llvm-native/Setup.hs [hunk 1]
```diff
@@ -1,2 +1,64 @@
...
+import System.FilePath
+
+
+main :: IO ()
+main = defaultMainWithHooks simpleUserHooks
+  { postConf = postConfHook
+  , preBuild = readHook buildVerbosity
+  }
+  where
+    readHook :: (a -> Flag Verbosity) -> Args -> a -> IO HookedBuildInfo
```
Comment (robbert-vdh) [inline accelerate-llvm-native/Setup.hs:30]: `Flag` is provided by both `Distribution.Simple.Setup` and `Distribution.PackageDescription`.

File: accelerate-llvm-native/src/Data/Array/Accelerate/LLVM/Native/Link/Runtime.hs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+-- Dynamic object loading
+-- ----------------------
+
+-- Load the shared object file and return pointers to the executable
+-- functions defined within
+--
+loadSharedObject :: HasCallStack => [ShortByteString] -> FilePath -> IO (FunctionTable, ObjectCode)
+loadSharedObject nms path = do
+  so      <- dlopen path []
```
Comment (robbert-vdh) [inline accelerate-llvm-native/src/Data/Array/Accelerate/LLVM/Native/Link/Runtime.hs:48]: Argument are missing for `dlopen()`. You need at least `RTLD_LAZY` so the symbols are lazily loaded, and you also want `RTLD_LOCAL` so the symbols defined in the kernel are not leaked to the global namespace (`RTLD_LOCAL` is the default, but it never hurts to be specific):

https://github.com/robbert-vdh/accelerate-llvm/blob/760ecd3822d1fd103783725067d4eb0de124511a/accelerate-llvm-native/src/Data/Array/Accelerate/LLVM/Native/Link/Posix.hs#L89

**Review by robbert-vdh — CHANGES_REQUESTED** [review]
I tried rebasing my `feature/tracy-annotations` branch (which contains the things needed for Tracy intergration, DWARF symbols, and the loop unrolling things) on this branch but after fixing the two things I linked the program just hung when trying to execute a kernel. Didn't have the time to investigate why though.

File: accelerate-llvm-native/src/Data/Array/Accelerate/LLVM/Native/Link/Runtime.hs (hunk 2 above)
Comment (tmcdonell) [reply accelerate-llvm-native/src/Data/Array/Accelerate/LLVM/Native/Link/Runtime.hs]: fixed

**Pull request merged**